plugins/events-node/src/api/EventRouter.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { EventBroker } from './EventBroker';
+import { EventSubscriber } from './EventSubscriber';
+import { EventPublisher } from './EventPublisher';
+import { EventParams } from './EventParams';
+
+/**
+ * Subscribes to the provided (generic) topic
+ * and publishes the events under the more concrete sub-topic
+ * depending on the implemented logic for detecting it.
+ * Implementing classes might use information from `metadata`
+ * and/or properties within the payload.
+ *
+ * Events of other topics will be ignored (should never happen).
+ *
+ * @public
+ */
+export abstract class EventRouter implements EventPublisher, EventSubscriber {
+  private eventBroker?: EventBroker;
+
+  protected constructor(private readonly topic: string) {}
+
+  protected abstract detectSubTopic(params: EventParams): string | undefined;
+
+  async onEvent(params: EventParams): Promise<void> {
+    if (params.topic !== this.topic) {
+      return;
+    }
+
+    const subTopic = this.detectSubTopic(params);
+    if (!subTopic) {
+      return;
+    }
+
+    this.eventBroker?.publish({
+      ...params,
+      topic: `${params.topic}/${subTopic}`,
```

Comment:
Currently, I use `/` as separate between topic and sub-topic. However, using this e.g. at an URL path might be a bit cumbersome. Maybe a different separator char like `.` (e.g., used by Azure DevOps) or `:` (e.g., used by BitbucketCloud) would make sense.

Reply:
```
foo/bar    =>   foo%2Fbar
foo:bar    =>   foo%3Abar
foo.bar    =>   foo.bar
foo_bar    =>   foo_bar
```

Azure DevOps uses `.` as separator, Bitbucket Cloud uses `:` as separator, GitHub uses `_` within names and as separator.

When used at Yaml files / the app-config, `.` could be misinterpreted as separator.

So far I tend to keeping `/` as separator within topic names.

Reply:
we use them as key at the `aws-sqs` module

from its `README.md`:

```yaml
events:
  modules:
    awsSqs:
      awsSqsConsumingEventPublisher:
        topics:
          topicName1: # replace with actual topic name as expected by subscribers
            queue:
              url: 'https://sqs.us-east-2.amazonaws.com/123456789012/MyQueue'
              region: us-east-2
              # visibilityTimeout - as HumanDuration; defaults to queue-based config
              # waitTime - as HumanDuration; defaults to max of 20 seconds (long polling)
            # timeout - as HumanDuration; timeout for the task execution
            # waitTimeAfterEmptyReceive - as HumanDuration; time to wait before a retry when there was no message.
          topicName2:
            # [...]
```

Reply:
alright, I decided spontaneously to switch to `.`. As the use of the topic name in configs is limited - esp. for cases where a sub-topic is used - it seems reasonable and you can still quote it at Yaml keys.